src/almanack/batch.py

```diff
@@ -0,0 +1,235 @@
+"""
+Batch processing utilities for running Almanack across many repositories.
+"""
+
+import json
+import logging
+from concurrent.futures import Executor, ProcessPoolExecutor, as_completed
+from pathlib import Path
+from typing import Any, Callable, List, Optional, Sequence, Type, Union
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+import pyarrow.parquet as pq
+
+from almanack.metrics.data import process_repo_for_almanack
+
+RepositoryProcessor = Callable[[str], dict[str, Any]]
+LOGGER = logging.getLogger(__name__)
+
+
+def sanitize_for_parquet(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Cleans a DataFrame so all columns are parquet-safe.
+
+    - Expands dict columns into multiple fields.
+    - Converts lists into JSON strings.
+    - Casts generic object types into strings.
+
+    Args:
+        df: Input DataFrame with raw metrics.
+
+    Returns:
+        df: Sanitized DataFrame safe for parquet storage.
+    """
+    for col in df.columns:
+        series = df[col]
+        # Check if column contains dicts
+        if series.apply(lambda x: isinstance(x, dict)).any():
+            nested = pd.json_normalize(series)
+            nested.columns = [f"{col}_{c}" for c in nested.columns]
+            df = df.drop(columns=[col]).join(nested)
+        # Check if column contains lists
+        elif series.apply(lambda x: isinstance(x, list)).any():
+            df[col] = series.apply(
+                lambda x: json.dumps(x) if isinstance(x, list) else x
+            )
+        # Fallback for generic objects
+        elif series.dtype == "object":
+            non_null = series.dropna()
+            is_numeric_like = (
+                not non_null.empty
+                and non_null.map(
+                    lambda x: isinstance(x, (int, float, bool, np.number))
+                ).all()
+            )
+            if is_numeric_like:
+                df[col] = pd.to_numeric(series, errors="coerce")
+            else:
+                # Preserve None as None; stringify only real values
+                df[col] = series.apply(lambda x: x if x is None else str(x))
+    return df
+
+
+def _nullable_dtype(dtype: Any) -> Any:
```

Comment:
this doesn't feel like a "batch" function, but I could be wrong

Reply:
This related to data fixes for batches of data from the Almanack. It might be able to fit somewhere else but this made the most sense at the time. I'll see if there's another spot. Generally I wanted to avoid creating a single function module if possible.